### ISSUE TITLE
what's the structure of AAlbert's classifier?

### ISSUE THREAD

**Issue by myhrbeu** [s3prl/s3prl#154]
Hi,
Now I want to reproduce the experimental results of AAlbert and Mockingjay, but the detailed structure of the classifier after feature extraction has not been introduced in detail in the paper. 
In the source code of Mockingjay, I found two classifier structures:
if self.model_type == 'linear':
            self.classifier = LinearClassifier(input_dim=input_dim,
                                               class_num=self.dataloader.dataset.class_num,
                                               task=self.task,
                                               dconfig=self.config['downstream']['linear'],
                                               sequencial=False).to(self.device)
        elif self.model_type == 'rnn':
            self.classifier = RnnClassifier(input_dim=input_dim,
                                            class_num=self.dataloader.dataset.class_num,
                                            task=self.task,
                                            dconfig=self.config['downstream']['rnn']).to(self.device)

The structure of AAlbert's classifier is only briefly introduced in the paper:
   1. To measure the phonetic information, we train 2-layer phoneme classifiers,
    2.Then, we obtain the force-aligned phoneme sequences, which contains 72 possible phone classes, with Montreal Forced Aligner
   3. Besides, the model here is a linear classifier with a mean-pooling layer.
  
But I still can't know the specific structure. Could you please tell me the specific structure in the paper of AAlbert, or where I can find it? Thank you!

**Comment by andi611** [comment]
Hi,

I can answer the questions regarding Mockingjay.

In the Mockingjay paper, a feed-forward classifier is used for phoneme classification.
Its implementation can be found [here](https://github.com/andi611/Mockingjay-Speech-Representation/blob/8f41f5728bdb94497e939fee0d67c7f65729a035/downstream/model.py#L21), and we use these [parameters](https://github.com/andi611/Mockingjay-Speech-Representation/blob/8f41f5728bdb94497e939fee0d67c7f65729a035/config/mockingjay_libri.yaml#L84-L88).

In a latter work, we present a **more standard setting for linear phoneme classification** of Mockingjay. This [setting](https://github.com/s3prl/s3prl/tree/master/s3prl/downstream/phone_linear) considers a [1-layer linear classifier](https://github.com/s3prl/s3prl/blob/ee5fd48b4f3e859ee90adc71e0907cd4eb380402/s3prl/downstream/phone_linear/model.py#L20), on a phone set of 41 possible phone classes (proposed and used in the [CPC paper](https://arxiv.org/pdf/1807.03748.pdf) written by DeepMind).

**We evaluate TERA, Mockingjay and also Audio-ALBERT with this linear phoneme classification setting** (Phoneme - LibriSpeech, linear, column):
![image](https://user-images.githubusercontent.com/25531495/126518669-7f33db1d-1897-428e-83c5-b82c6b674ed2.png)

You can also find the above result on [IEEE Early Access](https://ieeexplore.ieee.org/document/9478264) (The arXiv of TERA is not updated yet with the above results).

IMO this latter setting will be easier to reproduce, since we support this task in our repo.
If you consider this setting, I believe it will also resolves your Audio ALBERT related questions.

I hope this helps,
Andy

**Comment by toco2270853** [comment]
Dear researcher, It turns out that the result in IEEE Early Access is much higher than the one in Arxiv. Is there any important modification being made?

**Comment by andi611** [comment]
> Dear researcher, It turns out that the result in IEEE Early Access is much higher than the one in Arxiv. Is there any important modification being made?

Hi, in the IEEE Early Access version,
we change the input / pre-training features of TERA and Mockingjay from "fMLLR" to "80-dim log Mel", 
also, on the downstream tasks we tune the learning rate (which we did not in the Arxiv version) and perform a more thorough training.

**Comment by mn7ark** [comment]
Dear researcher, thanks for the quick reply.
  I also wonder how learning rate is set for different model Table 2. Did you just use a fix value, or search a best learning rate for different model thoroughly?

**Comment by andi611** [comment]
Hi, thanks for bringing up this question.

To clarify my previous response, I think changing from "fMLLR to 80-dim log Mel" and "perform a more thorough training" are the main contributors to performance increase, and not the learning rate tuning, as these tasks are not sensitive to the learning rate, below I will explain more.

At first, we tune the learning rate for every model, 
however, we later find that these tasks we adopt are not sensitive to upstream model variations.
So in a way, you can view it as we tune the learning rate for each downstream task (rather than the upstream model).

Empirically, we believe the results we report for each model are the optimal result, given that the downstream model architecture remains the same. Training all converge to similar testing performance, despite the seed/trial/and or even learning rate, if you give the training enough update steps (what I meant by "perform a more thorough training"). Overall these tasks are very stable for training.

The tasks we considered include `phone_linear`, `phone_1hidden`, `phone_linear_concat`, `timie_phone_linear`, `timit_phone_1hidden`, `timit_phone_linear_concat`, `speaker_linear_frame_libri`, and `speaker_linear_utter_libri`, which I have set their default learning rates to the ones I found generally working well. 

Andy

**Comment by mn7ark** [comment]
Thanks for your clear explanation!

**Comment by pohanchi** [comment]
Hi, sorry for the very late reply. I can answer the question for AAlbert.

The 2-layer phoneme classifiers is the same as Mockingjay paper. To clarify, The structure is [here](https://github.com/andi611/Mockingjay-Speech-Representation/blob/8f41f5728bdb94497e939fee0d67c7f65729a035/downstream/model.py#L21), which Andy is mentioned before.

Since we use different phoneme set for phoneme recognition tasks, so the result will be different. Please refer to Andy standard experiment. That will show more clear results to compare other models.

Then, linear classifier with a mean-pooling layer is for utterance-level speaker recognition task. The structure is [here](https://github.com/s3prl/s3prl/blob/b94e393acf9bf78cc3301cf64b18bf57b473173c/downstream/model.py#L500). Hope the explanation is clear.

PoHan
